docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,536 @@
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Proposed Solution
+In this solution, There will be updates in 
+1) **ACK Code Generator** to capture the late initialized fields from AWS operation output and use them to late initialize k8s custom resource object. Code generator changes will support a) configuration for service teams to tell ACK code generator which fields will be late initialized and b) configuration to provide custom code for late initialization.
+2) **ACK Runtime** to a) update the custom resource stored in etcd with late initalized optional spec fields And b) handle the late initalized fields for resources that have asynchronous creation.
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. Fix the latent bug where Spec fields are not set from the Create and Update response. Once this bug is fixed the late initialized fields which are present in Create and Update response will be set.
+    ```go
+    // Current Code
+    // sdk.go.tpl    
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 false }}
+        ...
+    }
+    ```
+
+    ```go
+    // Proposed Code
+    // sdk.go.tpl
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        // NOTE: The performSpecUpdate boolean param will be removed
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 }}
+        ...
+    }
+    ```
+
+2. **Provide a new "LateInitializationConfig" struct inside generator config for AWS service teams to provide which spec fields will be late initialized.** These fields will be the ones which are late initialized by the server but not present in the Create and Update response.
+
+    These fields will be populated by making the readOne call. AWS service teams can also provide a delay after which these fields will be late initialized.
+    LateInitializationConfig will be added as part of FieldConfig. 
+
+    ```go
+        // LateInitializeConfig contains instructions for how to handle the
+        // retrieval and setting of server-side defaulted fields.
+        type LateInitializationConfig struct {
+            // DelaySeconds provides the base delay to attempt late initialization again after an unsuccessful attempt
+            // to late initialized fields from ReadOne output
+            DelaySeconds int `json:"delay_seconds,omitempty"`
```

Comment:
nit: We should call this `MinDelaySeconds` (`min_delay_seconds`) since the reconciler logic will get the maximum across all fields, which means that most fields will be delayed by more than this amount.